### ISSUE TITLE
assignment 45

### ISSUE THREAD

**Issue by dipaavhad54-gif** [dipaavhad54-gif/python-assignments-#2]
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler

# =========================================================
# Create DataFrame
# =========================================================

data = {
    'Name': ['Amit', 'Sagar', 'Pooja'],
    'Math': [85, 90, 78],
    'Science': [92, 88, 80],
    'English': [75, 85, 82]
}

df = pd.DataFrame(data)

print("=" * 50)
print("ORIGINAL DATAFRAME")
print("=" * 50)
print(df)

# =========================================================
# Q1: Normalize Math scores using Min-Max Scaling
# =========================================================

print("\n" + "=" * 50)
print("Q1: MIN-MAX SCALING")
print("=" * 50)

scaler = MinMaxScaler()

df['Math'] = scaler.fit_transform(df[['Math']])

print(df)

# =========================================================
# Q2: Create Gender column and perform One-Hot Encoding
# =========================================================

print("\n" + "=" * 50)
print("Q2: GENDER AND ONE-HOT ENCODING")
print("=" * 50)

# Create Gender column
df['Gender'] = ['Male', 'Male', 'Female']

print("Before One-Hot Encoding:")
print(df)

# One-Hot Encoding
df = pd.get_dummies(df, columns=['Gender'], dtype=int)

print("\nAfter One-Hot Encoding:")
print(df)

# =========================================================
# Q3: Group students by gender and calculate average marks
# =========================================================

print("\n" + "=" * 50)
print("Q3: GROUP BY GENDER")
print("=" * 50)

# Original gender information पुन्हा वापरून grouping
gender_data = {
    'Name': ['Amit', 'Sagar', 'Pooja'],
    'Math': [85, 90, 78],
    'Science': [92, 88, 80],
    'English': [75, 85, 82],
    'Gender': ['Male', 'Male', 'Female']
}

gender_df = pd.DataFrame(gender_data)

average_marks = gender_df.groupby('Gender')[['Math', 'Science', 'English']].mean()

print(average_marks)

# =========================================================
# Q4: Pie chart of subject marks for Sagar
# =========================================================

print("\n" + "=" * 50)
print("Q4: SAGAR'S SUBJECT MARKS - PIE CHART")
print("=" * 50)

sagar = gender_df[gender_df['Name'] == 'Sagar'].iloc[0]

subjects = ['Math', 'Science', 'English']
marks = [sagar['Math'], sagar['Science'], sagar['English']]

plt.pie(marks, labels=subjects, autopct='%1.1f%%')
plt.title("Sagar's Subject Marks")
plt.show()

# =========================================================
# Q5: Add Status column
# Total >= 250 = Pass
# Otherwise = Fail
# =========================================================

print("\n" + "=" * 50)
print("Q5: PASS / FAIL STATUS")
print("=" * 50)

gender_df['Total'] = (
    gender_df['Math'] +
    gender_df['Science'] +
    gender_df['English']
)

gender_df['Status'] = gender_df['Total'].apply(
    lambda x: 'Pass' if x >= 250 else 'Fail'
)

print(gender_df)

# =========================================================
# Q6: Count how many students passed
# =========================================================

print("\n" + "=" * 50)
print("Q6: NUMBER OF PASSED STUDENTS")
print("=" * 50)

passed_count = (gender_df['Status'] == 'Pass').sum()

print("Number of students passed:", passed_count)

# =========================================================
# Q7: Export final DataFrame to CSV file
# =========================================================

print("\n" + "=" * 50)
print("Q7: EXPORT TO CSV")
print("=" * 50)

gender_df.to_csv("student_final.csv", index=False)

print("DataFrame exported successfully!")
print("File name: student_final.csv")

# =========================================================
# Q8: Histogram of Math marks
# =========================================================

print("\n" + "=" * 50)
print("Q8: HISTOGRAM OF MATH MARKS")
print("=" * 50)

plt.hist(gender_df['Math'], bins=5)

plt.xlabel("Math Marks")
plt.ylabel("Number of Students")
plt.title("Histogram of Math Marks")

plt.show()

# =========================================================
# Q9: Rename Math column to Mathematics
# =========================================================

print("\n" + "=" * 50)
print("Q9: RENAME MATH COLUMN")
print("=" * 50)

gender_df = gender_df.rename(
    columns={'Math': 'Mathematics'}
)

print(gender_df)

# =========================================================
# Q10: Boxplot for English marks
# =========================================================

print("\n" + "=" * 50)
print("Q10: BOXPLOT OF ENGLISH MARKS")
print("=" * 50)

plt.boxplot(gender_df['English'])

plt.ylabel("English Marks")
plt.title("Boxplot of English Marks")

plt.show()
